File: backend/ai/openai_provider.py
"""
Image-to-image: OpenAI images.edit (gpt-image) — фото комнаты или планировка → новый дизайн.
"""
import io
import os
import logging
from openai import OpenAI
from ai.base import AIProvider
from ai.config import PROMPT_MAX_CHARS, APARTMENT_IMAGES_COUNT
from ai.prompts import get_room_prompt, get_apartment_view_prompts
from ai.styles import STYLES, BUDGETS

log = logging.getLogger(__name__)
client = OpenAI(api_key=os.getenv("OPENAI_API_KEY"))


def _edit_image(image_bytes: bytes, prompt: str, filename: str = "room.jpg") -> str:
    """Image-to-image: редактирование фото/планировки по промпту."""
    prompt = prompt[:PROMPT_MAX_CHARS] if len(prompt) > PROMPT_MAX_CHARS else prompt
    f = io.BytesIO(image_bytes)
    f.name = filename
    result = client.images.edit(
        model="gpt-image-1.5",
        image=f,
        prompt=prompt,
    )
    d = result.data[0]
    if getattr(d, "url", None):
        return d.url
    if getattr(d, "b64_json", None):
        return f"data:image/png;base64,{d.b64_json}"
    raise ValueError("No image data in response")


def _generate_from_text(prompt: str) -> str:
    """Fallback: text-to-image, если image edit недоступен."""
    prompt = prompt[:PROMPT_MAX_CHARS] if len(prompt) > PROMPT_MAX_CHARS else prompt
    result = client.images.generate(
        model="dall-e-3",
        prompt=prompt,
        size="1024x1024",
        quality="standard",
        style="natural",
    )
    return (result.data[0].url or "").strip()


class OpenAIProvider(AIProvider):
    async def redesign_room(
        self,
        image_bytes: bytes,
        room_type: str,
        style: str,
        budget: str,
        user_text: str,
    ) -> str:
        """Image-to-image: фото комнаты → новый дизайн по стилю."""
        style_desc = STYLES.get(style, STYLES["minimalist"])
        budget_desc = BUDGETS.get(budget, BUDGETS["medium"])
        prompt = get_room_prompt(room_type, style_desc, budget_desc, user_text)
        full = f"Redesign this room. Keep the same layout and camera angle. {prompt} Photorealistic interior."
        return _edit_image(image_bytes, full)

    async def redesign_apartment(
        self,
        plan_image_bytes: bytes,
        user_preferences: str,
    ) -> list[str]:
        """Планировка: image-to-image — планировка → реалистичные фото по зонам."""
        view_prompts = get_apartment_view_prompts(user_preferences)[:APARTMENT_IMAGES_COUNT]
        images = []
        for i, view_prompt in enumerate(view_prompts):
            full = f"Based on this floor plan, generate a photorealistic interior photo. {view_prompt} Professional interior photography, natural lighting."
            images.append(_edit_image(plan_image_bytes, full, filename="plan.jpg"))
        return images

File: backend/ai/openai-provider.js
const OpenAI = require("openai");
const { File } = require("formdata-node");
const sharp = require("sharp");
const { getRoomPrompt, getApartmentViewPrompts } = require("./prompts");
const { STYLES, BUDGETS } = require("./styles");
const { PROMPT_MAX_CHARS, APARTMENT_IMAGES_COUNT } = require("./config");
const { toPngForApi, createMask } = require("../image-utils");

const client = new OpenAI({ apiKey: process.env.OPENAI_API_KEY });

const EDIT_MODEL = "dall-e-2";

function truncatePrompt(p, max) {
  return p.length > max ? p.slice(0, max) : p;
}

async function editImage(imagePngBuffer, prompt, filename = "room.png", maskSize = 1024) {
  const imageFile = new File([imagePngBuffer], filename, { type: "image/png" });
  const maskBuffer = await createMask(maskSize);
  const maskFile = new File([maskBuffer], "mask.png", { type: "image/png" });
  const data = await client.images.edit({
    model: EDIT_MODEL,
    image: imageFile,
    mask: maskFile,
    prompt: truncatePrompt(prompt, PROMPT_MAX_CHARS),
    size: "1024x1024",
  });
  const d = data.data[0];
  if (d?.url) return d.url;
  if (d?.b64_json) return `data:image/png;base64,${d.b64_json}`;
  throw new Error("No image data");
}

class OpenAIProvider {
  async redesignRoom(imageBytes, roomType, style, budget, userText, _mimetype = "") {
    const styleDesc = STYLES[style] || STYLES.minimalist;
    const budgetDesc = BUDGETS[budget] || BUDGETS.medium;
    const prompt = getRoomPrompt(roomType, styleDesc, budgetDesc, userText || "");
    const full = `Transform this room image into a ${styleDesc} style ${roomType}. ${prompt} Completely redesign: replace all furniture with new ${styleDesc} pieces, change wall colors, update flooring, add new decor and accessories. Transform the entire space while maintaining the same camera angle and room structure. Create a brand new interior design that matches the ${styleDesc} aesthetic.`;
    const png = await toPngForApi(imageBytes);
    const meta = await sharp(png).metadata();
    const size = meta.width || 1024;
    return editImage(png, full, "room.png", size);
  }

  async redesignApartment(planImageBytes, userPreferences, _mimetype = "") {
    const viewPrompts = getApartmentViewPrompts(userPreferences).slice(0, APARTMENT_IMAGES_COUNT);
    const png = await toPngForApi(planImageBytes);
    const meta = await sharp(png).metadata();
    const size = meta.width || 1024;
    const images = [];
    for (let i = 0; i < viewPrompts.length; i++) {
      const full = `Based on this floor plan, generate a photorealistic interior photo. ${viewPrompts[i]} Professional interior photography, natural lighting.`;
      const url = await editImage(png, full, `plan_${i}.png`, size);
      images.push(url);
    }
    return images;
  }

  async redesignApartmentSingle(planImageBytes, userPreferences, index, _mimetype = "") {
    const viewPrompts = getApartmentViewPrompts(userPreferences);
    if (index < 0 || index >= viewPrompts.length) {
      throw new Error(`Invalid index: ${index}. Must be between 0 and ${viewPrompts.length - 1}`);
    }
    const png = await toPngForApi(planImageBytes);
    const meta = await sharp(png).metadata();
    const size = meta.width || 1024;
    const full = `Based on this floor plan, generate a photorealistic interior photo. ${viewPrompts[index]} Professional interior photography, natural lighting.`;
    return editImage(png, full, `plan_${index}.png`, size);
  }
}

module.exports = OpenAIProvider;
